File: linkedlist.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LList:
    def __init__(self):
        self.size = 0
        self.head = None
        self.tail = None

    # Print the size
    def get_size(self):
        return self.size

    def is_empty(self):
        return self.size == 0
    
    def clear(self):
        self.head = self.tail = None
        self.size = 0

    def get_first(self):
        return self.head.data if self.head else None

    def get_last(self):
        return self.tail.data if self.tail else None

    # Print the linked list
    def print_list(self):
        cur = self.head
        while cur:
            print(cur.data, end=" ")
            cur = cur.next
        print()

    # Insert node at the beginning of the list
    def add_first(self, item):
        temp = Node(item)
        temp.next = self.head
        if self.head is None:
            self.tail = temp
        self.head = temp
        self.size += 1
    
    # Insert node before a specific node
    def add_before(self, next_node, item):
        if next_node is None:
            return
        temp = Node(item)
        if self.head == next_node:  # If next_node is head
            temp.next = self.head
            self.head = temp
        else:
            prev = self.head
            while prev and prev.next != next_node:
                prev = prev.next
            if prev:
                temp.next = prev.next
                prev.next = temp
        self.size += 1

    # Insert node after a specific node
    def add_after(self, prev, item):
        if prev is None:
            return
        temp = Node(item)
        temp.next = prev.next
        prev.next = temp
        if temp.next is None:  # If added at the end, update tail
            self.tail = temp
        self.size += 1

    # Get the node at a specific index
    def get_node_by_index(self, index):
        if index < 0:
            return None
        p = self.head
        for i in range(index):
            if p is None:
                return None
            p = p.next
        return p

    # Insert node at a specific index
    def add_by_index(self, index, item):
        if index < 0:
            return self.head

        if index == 0:
            self.add_first(item)
            return self.head

        prev = self.get_node_by_index(index-1)
        if prev:
            self.add_after(prev, item)
        self.size += 1
        return self.head

    # Insert into ordered list
    def add_to_ordered_list(self, item):
        p = self.head
        q = None

        while p and p.data < item:
            q = p
            p = p.next

        if q is None:
            self.add_first(item)
        else:
            self.add_after(q, item)
        

    # Find node of value
    def find_node(self, item):
        cur = self.head
        while cur:
            if cur.data == item:
                return cur
            cur = cur.next
        return None

    # Remove first node
    def remove_first(self):
        if self.head is None:
            return None
        temp = self.head
        self.head = self.head.next
        if self.head is None:
            self.tail = None
        self.size -= 1
        return self.head
    
    # Remove last node
    def remove_last(self):
        if self.tail is None:
            return None
        if self.head == self.tail:  # Only one element
            self.head = self.tail = None
        else:
            prev = self.head
            while prev.next != self.tail:
                prev = prev.next
            prev.next = None
            self.tail = prev
        self.size -= 1
        return self.tail

    # Remove node after a specific node
    def remove_after(self, prv):
        if prv is None or prv.next is None:
            return
        temp = prv.next
        prv.next = temp.next
        if prv.next is None:
            self.tail = prv
        self.size -= 1

    # Remove node by index
    def remove_by_index(self, index):
        if index < 0:
            return self.head

        if index == 0:
            return self.remove_first()

        prev = self.get_node_by_index(index-1)
        if prev and prev.next:
            self.remove_after(prev)
        self.size -= 1
        return self.head

    # Remove node by value
    def remove_by_value(self, item):
        p = self.head
        q = None

        while p and p.data != item:
            q = p
            p = p.next

        if p is None:
            return self.head

        if q is None:
            return self.remove_first()
        else:
            self.remove_after(q)
        return self.head

    # Function to merge two sorted lists
    @staticmethod
    def merge(list1, list2):
        dummy = Node(0)
        tail = dummy

        while list1 and list2:
            if list1.data <= list2.data:
                tail.next = list1
                list1 = list1.next
            else:
                tail.next = list2
                list2 = list2.next
            tail = tail.next

        tail.next = list1 if list1 else list2

        return dummy.next

    # Function to concatenate two lists
    def concat(self, list2):
        if self.head is None:
            self.head = list2.head
            self.tail = list2.tail
            self.size = list2.size
        elif list2.head is None:
            return
        else:
            self.tail.next = list2.head
            self.tail = list2.tail
            self.size += list2.size

    # Function to get the middle node
    def get_mid(self):
        slow = self.head
        fast = self.head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow

    # Function to sort the list (ascending order)
    def sort(self):
        if self.head is None or self.head.next is None:
            return self.head

        mid = self.get_mid()
        half = mid.next
        mid.next = None

        left_sorted = LList()
        left_sorted.head = self.head
        left_sorted.size = self.size // 2
        right_sorted = LList()
        right_sorted.head = half
        right_sorted.size = self.size - left_sorted.size

        left_sorted.head = left_sorted.sort()
        right_sorted.head = right_sorted.sort()
        self.head = self.merge(left_sorted.head, right_sorted.head)

        cur = self.head
        while cur and cur.next:
            cur = cur.next
        self.tail = cur

        return self.head
    
    def insertion_sort(self):
        if self.head is None or self.head.next is None:
            return
        
        sorted_list = None  # 새로운 정렬된 리스트의 시작 지점
        cur = self.head  # 정렬되지 않은 리스트의 노드를 탐색할 포인터
        
        while cur:
            next_node = cur.next  # 다음 노드를 미리 저장
            if sorted_list is None or sorted_list.data >= cur.data:
                cur.next = sorted_list
                sorted_list = cur
            else:
                temp = sorted_list
                while temp.next and temp.next.data < cur.data:
                    temp = temp.next
                cur.next = temp.next
                temp.next = cur
            
            cur = next_node  # 다음 노드로 이동
        
        self.head = sorted_list

    def selection_sort(self):
        if self.head is None or self.head.next is None:
            return
        
        cur = self.head
        
        while cur:
            min_node = cur
            r = cur.next
            
            while r:
                if r.data < min_node.data:
                    min_node = r
                r = r.next
            
            if min_node != cur:
                cur.data, min_node.data = min_node.data, cur.data
            
            cur = cur.next

    # Function to reverse the list
    def reverse(self):
        prev = None
        cur = self.head
        while cur:
            next_node = cur.next
            cur.next = prev
            prev = cur
            cur = next_node
        self.tail = self.head
        self.head = prev
        return self.head

    def rotate(self, d):
        """
        d%size: 실제로 회전하는 횟수(한바퀴, 즉 d == size인 경우는 제자리임으로)
        d = 0: 아무런 작업도 하지 않음
        d > 0: 오른쪽 회전
        d < 0: 왼쪽 회전
        """
        if self.head == None or d == 0 or self.size == 0:
            return
        d %= self.size #실제 회전수
        if d == 0: #그냥 한바퀴인 경우
            return
        if d < 0:
            d = self.size + d #왼쪽으로 d만큼이라는 것은 오른쪽으로 size-d만큼이라는것

        ##오른쪽 회전
        #1. 새로운 tail 먼저 지정
        new_tail = self.head
        for _ in range(self.size - d - 1):
            new_tail = new_tail.next

        
        #새로운 head: 새로운 tail 바로 옆
        new_head = new_tail.next
        new_tail.next = None #연결 끊기

        #원래 끊어진 부분 잇기
        self.tail.next = self.head

        #head, tail 갱신
        self.head = new_head
        self.tail = new_tail


    def hasCycle(self, head: Node) -> bool:
        slow_ptr = head
        fast_ptr = head
        
        while fast_ptr is not None and fast_ptr.next is not None:
            slow_ptr = slow_ptr.next
            fast_ptr = fast_ptr.next.next
            
            if slow_ptr == fast_ptr:
                return True
        
        return False

    def detectCycle(self, head: Node) -> Node:
        if head is None or head.next is None:
            return None
        
        slow_ptr = head
        fast_ptr = head
        entry_ptr = head
        
        while fast_ptr.next is not None and fast_ptr.next.next is not None:
            slow_ptr = slow_ptr.next
            fast_ptr = fast_ptr.next.next
            
            if slow_ptr == fast_ptr:
                while slow_ptr != entry_ptr:
                    slow_ptr = slow_ptr.next
                    entry_ptr = entry_ptr.next
                return entry_ptr
        
        return None
    
    
        # Convert linked list to array
    def linkedlist_to_array(self):
        arr = []
        current = self.head
        while current:
            arr.append(current.data)
            current = current.next
        return arr

    # Convert array to linked list
    def array_to_linkedlist(self, arr):
        self.clear()  # 기존 리스트를 초기화합니다.
        for item in arr:
            if self.tail is None:  # 리스트가 비어 있으면 첫 번째 요소로 추가
                self.add_first(item)
            else:
                self.add_after(self.tail, item)




if __name__ == "__main__":
    # Create an empty list and check if it's empty
    llist = LList()
    print("List initially empty:", llist.is_empty())

    # Add elements to the list
    llist.add_first(10)
    llist.add_first(20)
    llist.add_first(30)
    print("List after adding elements at the beginning:")
    llist.print_list()

    # Add elements by index
    llist.add_by_index(1, 15)
    llist.add_by_index(0, 25)
    print("List after adding elements by index:")
    llist.print_list()

    # Add to ordered list
    llist.add_to_ordered_list(18)
    llist.add_to_ordered_list(5)
    print("List after adding to ordered list:")
    llist.print_list()

    # Get first and last elements
    print("First element:", llist.get_first())
    print("Last element:", llist.get_last())

    # Find a node
    node = llist.find_node(15)
    print("Finding node with value 15:", "Found" if node else "Not found")

    # Add before a specific node
    node = llist.find_node(18)
    if node:
        llist.add_before(node, 17)
        print("List after adding 17 before 18:")
        llist.print_list()

    # Add after a specific node
    node = llist.find_node(10)
    if node:
        llist.add_after(node, 12)
        print("List after adding 12 after 10:")
        llist.print_list()

    # Remove elements
    llist.remove_by_value(25)
    print("List after removing element with value 25:")
    llist.print_list()

    llist.remove_by_index(2)
    print("List after removing element at index 2:")
    llist.print_list()

    llist.remove_first()
    print("List after removing the first element:")
    llist.print_list()

    llist.remove_last()
    print("List after removing the last element:")
    llist.print_list()

    # Remove after a specific node
    node = llist.find_node(15)
    if node:
        llist.remove_after(node)
        print("List after removing the node after 15:")
        llist.print_list()

    # Sort the list
    llist.selection_sort()
    print("List after selection sorting:")
    llist.print_list()

    # Reverse the list
    llist.reverse()
    print("List after reversing:")
    llist.print_list()

    # Rotate the list
    llist.rotate(2)
    print("List after rotating by 2 positions:")
    llist.print_list()

    # Sort the list
    llist.insertion_sort()
    print("List after insertion sorting:")
    llist.print_list()

    # Merge with another list
    list2 = LList()
    list2.add_first(40)
    list2.add_first(35)
    list2.add_first(28)
    print("Second list before merging:")
    list2.print_list()

    merged_head = LList.merge(llist.head, list2.head)
    merged_list = LList()
    merged_list.head = merged_head
    cur = merged_head
    size = 0
    while cur:
        size += 1
        cur = cur.next
    merged_list.size = size
    merged_list.tail = cur if cur else merged_list.get_node_by_index(size - 1)
    print("Merged list:")
    merged_list.print_list()

    # Concatenate another list
    list3 = LList()
    list3.add_first(55)
    list3.add_first(60)
    list3.add_first(65)
    print("Third list before concatenation:")
    list3.print_list()

    merged_list.concat(list3)
    print("List after concatenating third list:")
    merged_list.print_list()

    # Convert merged_list to array and back to linked list
    array_from_list = merged_list.linkedlist_to_array()
    print("Array after converting from linked list:", array_from_list)

    merged_list.array_to_linkedlist(array_from_list)
    print("Linked list after converting back from array:")
    merged_list.print_list()


    # Clear the list
    merged_list.clear()
    print("List after clearing:")
    print("Size of the list:", merged_list.get_size())
    merged_list.print_list()

File: linked_list/c/linkedlist.c
#include <stdio.h>
#include <stdlib.h>

typedef int element;

typedef struct Node {
    element data;
    struct Node* next;
} Node;

typedef struct LList {
    int size;
    Node *head;
    Node *tail;
} LList;

LList *create_list() {
    LList *list = (LList*)malloc(sizeof(LList));
    list->size = 0;
    list->head = NULL;
    list->tail = NULL;
    return list;
}

int get_size(LList* list) {
    return list->size;
}

int is_empty(LList* list) {
    return list->size == 0;
}

void clear(LList* list) {
    Node* current = list->head;
    Node* next_node;

    while (current != NULL) {
        next_node = current->next;
        free(current);
        current = next_node;
    }

    list->head = list->tail = NULL;
    list->size = 0;
}

int get_first(LList *list) {
    if (list->head != NULL) {
        return list->head->data;
    }
    return -1;
}

int get_last(LList * list) {
    if (list->tail != NULL) {
        return list->tail->data;
    }
    return -1;
}

// 리스트 출력
void print_list(LList *list) {
    Node *cur = list->head;
    while (cur != NULL) {
        printf("%d ", cur->data);
        cur = cur->next;
    }
    printf("\n");
}

// 연결리스트의 맨 앞에 삽입
void add_first(LList * list, element item) {
    Node *temp = (Node*)malloc(sizeof(Node));
    temp->data = item;
    temp->next = list->head;
    list->head = temp;
    if (list->tail == NULL) {
        list->tail = temp;
    }
    list->size++;
}

// 특정 노드 뒤에 삽입
void add_after(LList* list, Node* prev_node, int item) {
    if (prev_node == NULL) return;

    Node* temp = (Node*)malloc(sizeof(Node));
    temp->data = item;
    temp->next = prev_node->next;

    prev_node->next = temp;
    if (temp->next == NULL) {
        list->tail = temp;
    }
    list->size++;
}

// index에 해당하는 노드 찾기
Node* get_node_by_index(LList *list, int index) {
    if (index < 0) return NULL;
    Node *p = list->head;
    for (int i = 0; i < index && p != NULL; i++) {
        p = p->next;
    }
    return p;
}

// index에 해당하는 노드 삽입하기
void add_by_index(LList* list, int index, element item) {
    if (index < 0 || index > list->size) {
        return; // Handle out-of-bounds index
    }

    if (index == 0) {
        add_first(list, item);
        return;
    }

    Node *prev = get_node_by_index(list, index - 1);
    if (prev != NULL) {
        add_after(list, prev, item);
    }
}

// ordered list에 add하기
void add_to_ordered_list(LList *list, element item) {
    Node *p = list->head;
    Node *q = NULL;

    while (p != NULL && p->data < item) {
        q = p;
        p = p->next;
    }

    if (q == NULL) {
        add_first(list, item);
    } else {
        add_after(list, q, item);
    }
}

// 노드 찾기
Node* find_node(LList *list, element item) {
    Node *cur = list->head;
    while (cur != NULL) {
        if (cur->data == item) {
            return cur;
        }
        cur = cur->next;
    }
    return NULL;
}

// 첫번째 노드 지우기
void remove_first(LList *list) {
    if (list->head == NULL) return;
    Node *temp = list->head;
    list->head = list->head->next;
    free(temp);
    if (list->head == NULL) {
        list->tail = NULL;
    }
    list->size--;
}

// 마지막 노드 지우기
void remove_last(LList *list) {
    if (list->head == NULL) return;

    if (list->head == list->tail) {
        free(list->head);
        list->head = list->tail = NULL;
    } else {
        Node *temp = list->head;
        while (temp->next != list->tail) {
            temp = temp->next;
        }
        free(list->tail);
        list->tail = temp;
        list->tail->next = NULL;
    }
    list->size--;
}

// 해당 노드 다음 노드 지우기
void remove_after(LList* list, Node* prev_node) {
    if (prev_node == NULL || prev_node->next == NULL) return;

    Node* temp = prev_node->next;
    prev_node->next = temp->next;

    if (temp == list->tail) {
        list->tail = prev_node;
    }

    free(temp);
    list->size--;
}

// index로 노드 지우기
void remove_by_index(LList* list, int index) {
    if (index < 0) {
        return;
    }

    if (index == 0) {
        remove_first(list);
        return;
    }

    Node *prev = get_node_by_index(list, index - 1);
    if (prev != NULL && prev->next != NULL) {
        remove_after(list, prev);
    }
}

// 주어진 data에 해당하는 노드 지우기
void remove_by_value(LList* list, element item) {
    if (list->head == NULL) return;

    if (list->head->data == item) {
        remove_first(list);
        return;
    }

    Node* prev = list->head;
    Node* cur = list->head->next;

    while (cur != NULL && cur->data != item) {
        prev = cur;
        cur = cur->next;
    }

    if (cur != NULL) {
        remove_after(list, prev);
    }
}

// Function to merge two sorted lists
Node* merge(Node *list1, Node *list2) {
    Node dummy;
    Node *tail = &dummy;
    dummy.next = NULL;

    while (list1 != NULL && list2 != NULL) {
        if (list1->data <= list2->data) {
            tail->next = list1;
            list1 = list1->next;
        } else {
            tail->next = list2;
            list2 = list2->next;
        }
        tail = tail->next;
    }
    tail->next = (list1 != NULL) ? list1 : list2;
    return dummy.next;
}

// Function to concatenate two lists
void concat(LList* list1, LList* list2) {
    if (list1->head == NULL) {
        list1->head = list2->head;
        list1->tail = list2->tail;
    } else if (list2->head != NULL) {
        list1->tail->next = list2->head;
        list1->tail = list2->tail;
    }
    list1->size += list2->size;
    list2->head = list2->tail = NULL;
    list2->size = 0;
}

// Function to get the middle node
Node* get_mid(LList* list) {
    if (list == NULL || list->head == NULL) return NULL;
    Node* slow = list->head;
    Node* fast = list->head;

    while (fast->next != NULL && fast->next->next != NULL) {
        slow = slow->next;
        fast = fast->next->next;
    }
    return slow;
}

// Function to sort the list (오름차순 정렬)
void sort(LList *list) {
    if (list == NULL || list->head == NULL || list->head->next == NULL) {
        return; // Base case for recursion
    }

    Node *mid = get_mid(list);
    if (mid == NULL) return; // Safety check

    Node *half = mid->next;
    mid->next = NULL; // Split the list into two halves

    LList* left_sorted = create_list();
    LList* right_sorted = create_list();

    left_sorted->head = list->head;
    left_sorted->tail = mid;
    left_sorted->size = list->size / 2;

    right_sorted->head = half;
    right_sorted->tail = list->tail;
    right_sorted->size = list->size - left_sorted->size;

    sort(left_sorted);
    sort(right_sorted);

    list->head = merge(left_sorted->head, right_sorted->head);

    // Update tail pointer after merging
    Node* cur = list->head;
    while (cur && cur->next) {
        cur = cur->next;
    }
    list->tail = cur;
    list->size = left_sorted->size + right_sorted->size;

    free(left_sorted);
    free(right_sorted);
}

// Insertion Sort
void insertion_sort(LList* list) {
    if (list->head == NULL || list->head->next == NULL) {
        return;  // 리스트가 비어있거나 요소가 하나인 경우
    }

    Node* sorted = NULL;  // 정렬된 리스트의 시작점
    Node* cur = list->head;

    while (cur != NULL) {
        Node* next = cur->next;  // 다음 노드를 미리 저장
        cur->next = NULL;  // 현재 노드를 분리

        if (sorted == NULL || sorted->data >= cur->data) {
            cur->next = sorted;
            sorted = cur;
        } else {
            Node* temp = sorted;
            while (temp->next != NULL && temp->next->data < cur->data) {
                temp = temp->next;
            }
            cur->next = temp->next;
            temp->next = cur;
        }
        cur = next;  // 다음 노드로 이동
    }

    list->head = sorted;

    Node* end = sorted;
    while (end->next != NULL) {
        end = end->next;
    }
    list->tail = end;
}

// Selection Sort
void selection_sort(LList* list) {
    if (list->head == NULL || list->head->next == NULL) {
        return;  // 리스트가 비어있거나 요소가 하나인 경우
    }

    Node* cur = list->head;

    while (cur != NULL) {
        Node* min = cur;
        Node* r = cur->next;

        // 현재 위치에서 최소값을 찾음
        while (r != NULL) {
            if (r->data < min->data) {
                min = r;
            }
            r = r->next;
        }

        // 최소값을 현재 위치의 값과 교환
        if (min != cur) {
            element temp = cur->data;
            cur->data = min->data;
            min->data = temp;
        }

        cur = cur->next;  // 다음 노드로 이동
    }
}


// Function to reverse the list
void reverse(LList* list) {
    Node* prev = NULL;
    Node* cur = list->head;
    Node* next = NULL;

    while (cur) {
        next = cur->next;
        cur->next = prev;
        prev = cur;
        cur = next;
    }

    list->tail = list->head;
    list->head = prev;
}

// Rotate the list
void rotate(LList* list, int d) {
    if (list->head == NULL || d == 0 || list->size == 0) return;

    d = d % list->size;
    if (d == 0) return;

    if (d < 0) d = list->size + d;

    Node* new_tail = list->head;
    for (int i = 0; i < list->size - d - 1; i++) {
        new_tail = new_tail->next;
    }

    Node* new_head = new_tail->next;
    new_tail->next = NULL;

    Node* temp = list->head;
    list->head = new_head;

    Node* end = list->head;
    while (end->next != NULL) {
        end = end->next;
    }
    end->next = temp;
    list->tail = new_tail;
}

#include <stdbool.h>

bool hasCycle(struct Node *head) {
    struct Node *slow_ptr = head;
    struct Node *fast_ptr = head;
    while (fast_ptr != NULL && fast_ptr->next != NULL){
        slow_ptr = slow_ptr->next;
        fast_ptr = fast_ptr->next->next;
        if(slow_ptr == fast_ptr){
            return true;
        }
    }
    return false;
}

struct Node *detectCycle(struct Node *head) {
    
    // Initial Condition for for no node or no cycle
    if(head == NULL || head->next == NULL)return NULL;
    
    struct Node* slowptr = head;//Slow Pointer
    struct Node* fastptr = head;//Fast Pointer
    struct Node* entryptr = head; // entry location
    
    //Search of the loop cycle using fast pointer
    while(fastptr->next && fastptr->next->next){
        slowptr = slowptr->next;
        fastptr = fastptr->next->next;
        
        //Condition if cycle exist, if cycle exist then check the entry point of the cycle further
        if(slowptr == fastptr){
            
            //loop to check the entrypoint of the cycle, 
            //once found return the pointer to the entry point 
            while(slowptr != entryptr){
                slowptr = slowptr->next;
                entryptr = entryptr->next;
            }
            return entryptr;
        }
    }
    return NULL;
}

int* linkedlist_to_array(LList* list) {
    int* arr = (int*)malloc(list->size * sizeof(int));
    Node* current = list->head;
    int index = 0;
    while (current != NULL) {
        arr[index++] = current->data;
        current = current->next;
    }
    return arr;
}

void array_to_linkedlist(LList* list, int* arr, int array_size) {
    clear(list);  // 기존 리스트를 초기화합니다.
    for (int i = 0; i < array_size; i++) {
        if (list->head == NULL) {
            add_first(list, arr[i]);  // 리스트가 비어 있을 경우 첫 번째 요소 추가
        } else {
            add_after(list, list->tail, arr[i]);  // 그 이후 요소들은 tail 뒤에 추가
        }
    }
}


int main() {
    LList* llist = create_list();

    // Initial list setup
    printf("Adding 3, 2, 1 to the list using add_first:\n");
    add_first(llist, 3); // Thursday
    add_first(llist, 2); // Tuesday
    add_first(llist, 1); // Monday
    print_list(llist);

    // Get size of the list
    printf("Size of the list: %d\n", get_size(llist));

    // Get first and last elements
    printf("First element: %d\n", get_first(llist));
    printf("Last element: %d\n", get_last(llist));

    // Find a node
    printf("Finding node with value 2:\n");
    Node* found = find_node(llist, 2);
    printf("Found: %d\n", found != NULL ? found->data : -1);

    // Insert at index
    printf("Inserting 5 at index 1:\n");
    add_by_index(llist, 1, 5);
    print_list(llist);

    // Ordered insertion
    printf("Adding 4 to ordered list:\n");
    add_to_ordered_list(llist, 4);
    print_list(llist);

    // Reverse the list
    printf("Reversing the list:\n");
    reverse(llist);
    print_list(llist);

    // Sort the list
    printf("Insertion sort:\n");
    insertion_sort(llist);
    print_list(llist);

    // Rotate the list
    printf("Rotating the list by 2 positions:\n");
    rotate(llist, 2);
    print_list(llist);

    printf("Selection sorting:\n");
    selection_sort(llist);
    print_list(llist);

    // Remove first and last
    printf("Removing first element:\n");
    remove_first(llist);
    print_list(llist);

    printf("Removing last element:\n");
    remove_last(llist);
    print_list(llist);

    // Remove by value
    printf("Removing node with value 4:\n");
    remove_by_value(llist, 4);
    print_list(llist);

    // Clear the list
    printf("Clearing the list:\n");
    clear(llist);
    print_list(llist);

    // Creating two lists for merging and concatenating
    LList* list1 = create_list();
    LList* list2 = create_list();
    add_first(list1, 4);
    add_first(list1, 2);
    add_first(list1, 1);
    add_first(list2, 5);
    add_first(list2, 3);
    printf("List 1:\n");
    print_list(list1);
    printf("List 2:\n");
    print_list(list2);

    // Concatenate two lists
    printf("Concatenating list 1 and list 2:\n");
    concat(list1, list2);
    print_list(list1);

    // Convert list1 to array and back to linked list
    int* array_from_list = linkedlist_to_array(list1);
    printf("Array after converting from linked list:\n");
    for (int i = 0; i < get_size(list1); i++) {
        printf("%d ", array_from_list[i]);
    }
    printf("\n");

    array_to_linkedlist(list1, array_from_list, get_size(list1));
    printf("Linked list after converting back from array:\n");
    print_list(list1);

    free(array_from_list);  // 동적 할당한 배열 해제

    return 0;
}
